### PR TITLE
Update dark/light theme toggle

### LINKED ISSUE #721: Update the dark mode switch by adding a toggle switch button
### Context
Update the dark mode switch by adding a toggle switch buttong

### Screenshots are attached
this seems like an outdated way to switch to dark mode so i would like to enhance it.
Screenshot 2023-03-30 080904

enhance it to something like this

Screenshot 2023-03-30 081034

### If the feature request is approved, would you be willing to submit a PR?

Yes

### PR DESCRIPTION
**Pull request by Shubham-Patel07**
### What kind of changes does this PR include?

-   [x] Fixes or refactors: closes #721 
-   [ ] A new challenge
-   [ ] Additional documentation
-   [ ] Something else

#### Description
New enhancement done, Updated the radiio button based toggle switch to checkbox based toggle switch

from

228714471-d7090e59-a36e-4f45-9211-037ba0b8fed3

updated one,

291900265-fdc20c9c-03f0-4069-b7eb-21807f62dc9f

291900167-272802da-daf5-4b44-b1ce-22612c098d18

### CODE REVIEW

File: src/main/resources/templates/fragments/header.html [hunk 1]
```diff
@@ -51,16 +51,20 @@
...
-                    <div class="theme-toggle navbar-text">
-                        <label>
-                            <span class="visually-hidden">Light</span>
-                            <input type="radio" value="light" name="theme">
-                            <span class="checkmark" th:attr="data-cy='light-mode-radio'">☀️</span>
-                        </label>
-                        <label>
-                            <span class="visually-hidden">Dark</span>
-                            <input type="radio" value="dark" name="theme">
-                            <span class="checkmark" th:attr="data-cy='dark-mode-radio'">🌜</span>
```
Comment (commjoen) [inline src/main/resources/templates/fragments/header.html:63]: the attributes are required for people who have trouble seeing and are required for the cypress tests to run. Can you add the data-cy attrs to your new implementation please?
And the cypress tests need updating to use the new labels to be able to switch themes

File: src/main/resources/static/theme-toggle.js [hunk 2]
```diff
@@ -1,34 +1,38 @@
 (function () {
   const darkModeMediaQuery = window.matchMedia('(prefers-color-scheme: dark)')
-  let initialTheme
-  if (localStorage.getItem('darkmode-pref-set') === 'true') {
-    initialTheme = localStorage.getItem('darkMode') === 'true'
-  } else {
-    initialTheme = darkModeMediaQuery.matches
-  }
```
Comment (commjoen) [inline src/main/resources/static/theme-toggle.js:8]: can we prevent the switch from "animating" every time we load a page in dark mode please? this makes the rendering time quite lengthy . I rather only have the animation when somethiing changes by the user, but not on every page.
This is also one of the reasons why the cypress tests are failing.

**Review by commjoen — CHANGES_REQUESTED** [review]
Hi @Shubham-Patel07 , the icons look awesome! There are just a few things that will need some move love and then we are good to go, can you please:
- [ ] make sure the toggle only renders a switch when clicked/tapped
- [ ] make sure the attributes are back so the cypress tests can run (and/or adapt the cypress test for the toggle switching)

Cheers!

File: src/main/resources/static/theme-toggle.js (hunk 2 above)
Comment (Shubham-Patel07) [reply src/main/resources/static/theme-toggle.js:8]: i tried but i am not able to resolve that flickering issue every time we navigate to different page, but about the cypress test its failing cause they are written as respect to radio button logic and i have updated to checkbox hence they are failing. so i am tweaking the test script a bit

Comment (commjoen) [reply src/main/resources/static/theme-toggle.js:8]: the flickering issue: might that have to do with how you load the initial state and how you render it ;-)?
Happy to review again when you have the tests fixed and the attributes `th:attr="data-cy='dark-mode-<widgettype>'"` back again.

File: src/main/java/org/owasp/wrongsecrets/challenges/ChallengeUI.java [hunk 3]
```diff
@@ -105,7 +105,7 @@ private String documentation(Function<ChallengeSource, String> extractor) {
       return challengeDefinition.source(runtimeEnvironment).map(extractor).orElse("");
     } else {
       // We cannot run the challenge but showing documentation should still be possible
-      return extractor.apply(challengeDefinition.sources().getFirst());
+      return extractor.apply(challengeDefinition.sources().get(0));
```
Comment (commjoen) [inline src/main/java/org/owasp/wrongsecrets/challenges/ChallengeUI.java:108]: can you please undo the removal of `getFirst()`?

File: src/main/java/org/owasp/wrongsecrets/challenges/ChallengesController.java [hunk 4]
```diff
@@ -97,7 +97,7 @@ public String spoiler(@PathVariable("short-name") String shortName, Model model)
       Supplier<Spoiler> spoilerFromRandomChallenge =
           () -> {
             var challengeDefinition = findByShortName(shortName);
-            return challenges.getChallenge(challengeDefinition).getFirst().spoiler();
+            return challenges.getChallenge(challengeDefinition).get(0).spoiler();
```
Comment (commjoen) [inline src/main/java/org/owasp/wrongsecrets/challenges/ChallengesController.java:100]: can you please undo the removal of `getFirst()`?

Comment (Shubham-Patel07) [reply src/main/java/org/owasp/wrongsecrets/challenges/ChallengesController.java]: ya i will and i am pushing the changed test scripts right now, extremely sorry for the delay

Comment (Shubham-Patel07) [reply src/main/java/org/owasp/wrongsecrets/challenges/ChallengesController.java]: now you can look into it

File: src/main/resources/static/theme-toggle.js (hunk 2 above)
Comment (commjoen) [reply src/main/resources/static/theme-toggle.js:8]: cypress seems to work, i guess there are 2 things left:
- In the console of the browser, I now get an error: "Error: <path> attribute d: Expected number, "…7.18492 13.0678 ..."."
- it seems that when we set it to "checked" in the dom it only animates when we want to go to light mode. @mikewoudenberg would you know what to do here?

**Comment by Shubham-Patel07** [comment]
I have pushed the test scripts please look into flickering issue

**Comment by commjoen** [comment]
I have looked into it, but could not find a solution. Will have to postpone that for a while unless @bendehaan finds a solution maybe. I am also not sure if this control is usable for the people with visibility issues.

**Pull request closed without merging**